stats/tally.go

```diff
@@ -0,0 +1,188 @@
+package stats
+
+import (
+	"sync"
+	"time"
+
+	"github.com/uber/tchannel-go"
+
+	"github.com/uber-go/tally"
+)
+
+type wrapper struct {
+	sync.RWMutex
+
+	scope  tally.Scope
+	byTags map[knownTags]*taggedScope
+}
+
+type knownTags struct {
+	dest       string
+	source     string
+	procedure  string
+	retryCount string
+}
+
+type taggedScope struct {
+	sync.RWMutex
+
+	scope tally.Scope // already tagged with some set of tags
+
+	counters map[string]tally.Counter
+	gauges   map[string]tally.Gauge
+	timers   map[string]tally.Timer
+}
+
+// NewTallyReporter takes a tally.Scope and wraps it so it ca be used as a
+// StatsReporter. The list of metrics emitted is documented on:
+// https://tchannel.readthedocs.io/en/latest/metrics/
+// The tags emitted are similar to Muttley/YARPC, the tags emitted are:
+// source, dest, procedure, and retry-count.
+func NewTallyReporter(scope tally.Scope) tchannel.StatsReporter {
+	return &wrapper{
+		scope:  scope,
+		byTags: make(map[knownTags]*taggedScope),
+	}
+}
+
+func (w *wrapper) IncCounter(name string, tags map[string]string, value int64) {
+	ts := w.getTaggedScope(tags)
+	ts.getCounter(name).Inc(value)
+}
+
+func (w *wrapper) UpdateGauge(name string, tags map[string]string, value int64) {
+	ts := w.getTaggedScope(tags)
+	ts.getGauge(name).Update(float64(value))
+}
+
+func (w *wrapper) RecordTimer(name string, tags map[string]string, d time.Duration) {
+	ts := w.getTaggedScope(tags)
+	ts.getTimer(name).Record(d)
+}
+
+func (w *wrapper) getTaggedScope(tags map[string]string) *taggedScope {
+	kt := convertTags(tags)
+
+	w.RLock()
+	ts, ok := w.byTags[kt]
```

Comment:
In YARPC’s observability layer, we often digest the tags into a null-delimited string. You know the performance of struct key map lookup better than me. This is on the hot path?

Reply:
Good question -- this is definitely on the hot path (since we emit multiple stats per RPC).

A single string lookup is faster than using a struct with 4 strings. Just the lookup has a huge performance difference (the struct has 4 strings, and the string is the 4 strings concatenated with a null byte separator):
```
string-lookup: 24.6 ns/op
struct-lookup: 84.4 ns/op
```

However, this is not enough to look at, as using a single string requires building that string. Let's add the cost of building a string (without allocating) by reusing a shared buffer, and then do the lookup:
```
string-build-lookup: 63.8
```

This may be called from multiple goroutines, so we can't reuse a single buffer. Once we add buffer pooling, the cost difference is negligible, and the complexity of pooling (and risk of issues) is probably not worth the optimization.